[POAE7-2677] HashTable implemented with int8 and int16

Deegue opened the pull request:
### What changes were proposed in this pull request?
Support HashTable for key type of int8 and int16.

1.`FixedHashMap` is ready, which is designed for int8 and int16.
2.Added `CiderNewHashTableTest.cpp` and validated the usage of `FixedHashMap`.
3.Use `CiderDefaultAllocator` as default allocator for memory usage in HashTable.


### Why are the changes needed?
Implement new HashTables.


### Does this PR introduce _any_ user-facing change?
No.


### How was this patch tested?
UTs.

### Which label does this PR belong to?

Deegue commented:
APIs for codegen will be submitted in another pr.

Trying to add an exclusion file for license check.

It seems something wrong with code-style-check, there is no difference between the original version and the auto revised version.

jikunshang commented on cider/common/Stopwatch.h:
```diff
@@ -0,0 +1,198 @@
+ * under the License.
+ */
+
+#pragma once
```
necessary file?

Deegue replied:
It will be used in HashTable.h to record time cost of resize(rehash) when debug is enabled. Anyway, let me remove this since it's our first version and add back later.

jikunshang commented on cider/common/hashtable/FixedHashMap.h:
```diff
@@ -0,0 +1,189 @@
+};
+
+template <typename Key, typename Mapped, typename Allocator = HashTableAllocator>
+using FixedImplicitZeroHashMap =
```
`FixedImplicitZeroHashMap` `FixedImplicitZeroHashMapWithCalculatedSize` `FixedHashMap` are all interface you will provide? what's the difference and how to choose. Do we need ut for these basic class and API? Please provide more info via comments.

Deegue replied:
Yes, `FixedImplicitZeroHashMap` is for int8 and `FixedImplicitZeroHashMapWithCalculatedSize` is for int16. Let me add some comments here.

jikunshang commented on cider/common/hashtable/FixedHashSet.h:
```diff
@@ -0,0 +1,51 @@
+#include <common/hashtable/FixedHashTable.h>
+
+template <typename Key, typename Allocator = HashTableAllocator>
+class FixedHashSet
```
do we need this class?

Deegue replied:
Yes, it may be used in `Aggregator` to store all inserted keys.

jikunshang replied:
Please provide some examples for how to use `Aggregator` if necessary

Deegue replied:
Examples will be available in PR for [POAE7-2678](https://jira.devtools.intel.com/browse/POAE7-2678)

jikunshang commented on cider/tests/CiderNewHashTableTest.cpp:
```diff
@@ -0,0 +1,265 @@
+ protected:
+  static MockTable* MockTableForTest;
+
+  static void SetUpTestSuite() {
```
do we need this?

Deegue replied:
It will be used later, let me remove it now and add back in the future.

jikunshang commented on cider/tests/CiderNewHashTableTest.cpp:
```diff
@@ -0,0 +1,265 @@
+  bool operator!() const { return !count; }
+};
+
+TEST_F(CiderNewHashTableTest, aggInt8Test) {
```
I think this test is not verifying hashmap functionality, it's more like an example about how to do aggregation based on this hashmap.

Deegue replied:
Yes, these are examples of using `FixedImplicitZeroHashMap` and `FixedImplicitZeroHashMapWithCalculatedSize`.

jikunshang commented:
```
1.FixedHashMap is ready, which is designed for int8 and int16.
2.Added CiderNewHashTableTest.cpp and validated the usage of FixedHashMap.
```
I don't think you have verified all APIs `FixedHashMap` provided. 
And item 3456 are TODOs.

spevenhe commented:
Can we rename these class adding ck in their name or change the dir name to ck to indicate where they are from? Right now they are using basic names and may be confusing when we migrate more hashtables from other frameworks.

jikunshang commented on cider/common/Arena.h:
```diff
@@ -0,0 +1,122 @@
+ * - memory is allocated and freed by large MemoryChunks;
+ * - freeing parts of data is not possible (but look at ArenaWithFreeLists if you need);
+ */
+class Arena : private boost::noncopyable {
```
do we need an ArenaAllocator like https://github.com/duckdb/duckdb/blob/master/src/include/duckdb/storage/arena_allocator.hpp

Deegue replied:
Something like that but here it will only be used for string. Added some todos here for further implementation.

Deegue commented:
> Can we rename these class adding ck in their name or change the dir name to ck to indicate where they are from? Right now they are using basic names and may be confusing when we migrate more hashtables from other frameworks.

I think we don't need to do that since it's not a copy of code but picked useful parts. Later if we integrate other HTs, the first thing is to find out whether the code we wanna pick is logically existed in our repo or not. Meanwhile, we can also refactor the code if it is necessary.

spevenhe commented:
> > Can we rename these class adding ck in their name or change the dir name to ck to indicate where they are from? Right now they are using basic names and may be confusing when we migrate more hashtables from other frameworks.
> 
> I think we don't need to do that since it's not a copy of code but picked useful parts. Later if we integrate other HTs, the first thing is to find out whether the code we wanna pick is logically existed in our repo or not. Meanwhile, we can also refactor the code if it is necessary.

In this way, we still need to rename the class. Since we will have many hashtables, only the base interface class can be called ```hashtable``` and if other implemented class are also called ```hashtable``` , their name will be confusing.

Deegue commented:
> > > Can we rename these class adding ck in their name or change the dir name to ck to indicate where they are from? Right now they are using basic names and may be confusing when we migrate more hashtables from other frameworks.
> > 
> > 
> > I think we don't need to do that since it's not a copy of code but picked useful parts. Later if we integrate other HTs, the first thing is to find out whether the code we wanna pick is logically existed in our repo or not. Meanwhile, we can also refactor the code if it is necessary.
> 
> In this way, we still need to rename the class. Since we will have many hashtables, only the base interface class can be called `hashtable` and if other implemented class are also called `hashtable` , their name will be confusing.

`HashTable` is a basic interface class, it's universal and it won't be changed. Upper implementations may have different names like `FixedHashTable`.

spevenhe commented:
> > > > Can we rename these class adding ck in their name or change the dir name to ck to indicate where they are from? Right now they are using basic names and may be confusing when we migrate more hashtables from other frameworks.
> > > 
> > > 
> > > I think we don't need to do that since it's not a copy of code but picked useful parts. Later if we integrate other HTs, the first thing is to find out whether the code we wanna pick is logically existed in our repo or not. Meanwhile, we can also refactor the code if it is necessary.
> > 
> > 
> > In this way, we still need to rename the class. Since we will have many hashtables, only the base interface class can be called `hashtable` and if other implemented class are also called `hashtable` , their name will be confusing.
> 
> `HashTable` is a basic interface class, it's universal and it won't be changed. Upper implementations may have different names like `FixedHashTable`.

Yes. But in this PR, the ```hashtable```class in ```hashtable.h``` is not the basic interface class we descripted in MIP. It is more like an implementation. So I am just thinking its name is confusing.

Deegue commented:
> 



> > > > > Can we rename these class adding ck in their name or change the dir name to ck to indicate where they are from? Right now they are using basic names and may be confusing when we migrate more hashtables from other frameworks.
> > > > 
> > > > 
> > > > I think we don't need to do that since it's not a copy of code but picked useful parts. Later if we integrate other HTs, the first thing is to find out whether the code we wanna pick is logically existed in our repo or not. Meanwhile, we can also refactor the code if it is necessary.
> > > 
> > > 
> > > In this way, we still need to rename the class. Since we will have many hashtables, only the base interface class can be called `hashtable` and if other implemented class are also called `hashtable` , their name will be confusing.
> > 
> > 
> > `HashTable` is a basic interface class, it's universal and it won't be changed. Upper implementations may have different names like `FixedHashTable`.
> 
> Yes. But in this PR, the `hashtable`class in `hashtable.h` is not the basic interface class we descripted in MIP. It is more like an implementation. So I am just thinking its name is confusing.

`HashTable.h` is the basic interface class. APIs defined in our MIP changed recently and I have aligned `HashTable.h` to it.

chaojun-zhang commented on cider/common/Arena.h:
```diff
@@ -0,0 +1,122 @@
+
+#include <boost/noncopyable.hpp>
+
+namespace DB {
```
why use DB namespace?

Deegue replied:
Removed ns since it's useless currently.

Deegue commented:
Gentle ping @jikunshang @spevenhe @chaojun-zhang for another review, thanks!

jikunshang commented:
Please add some test case for hash table API. I don't think you have any case to cover emplace/find/etc. methods.
BTW, please try to rewrite PR descirption.

Deegue commented:
> Please add some test case for hash table API. I don't think you have any case to cover emplace/find/etc. methods. BTW, please try to rewrite PR descirption.

Added api test, rewrote change 3 of PR descirption.

jikunshang commented on src/cider/tests/CiderNewHashTableTest.cpp:
```diff
@@ -0,0 +1,216 @@
+
+TEST_F(CiderNewHashTableTest, hashTableTest) {
+  using Cell = FixedHashMapCell<int16_t, Block>;
+  FixedHashTable<int16_t, Cell, FixedHashTableStoredSize<Cell>, HashTableAllocator> fht;
```
the second parameter have to be `FixedHashMapCell`, possible to be some simple type for this ut/example?

yma11 replied:
@Deegue please address this comment in later PR.

Deegue replied:
Users will not need to care about Cell like `FixedHashMapCell<key_type, value_type>` or `HashMapCell<key_type, value_type>` since it's related to the HashTable implementation decided by the chooser. Users only need to pass down `key_type` and `value_type`.

jikunshang replied:
what should I do if I want a simple hash map which `key_type` is Int32 and `value_type` is Int64

Deegue replied:
When using `Aggregator`, you can set via constructor. But for hash map, it cannot avoid setting a `Cell`. It's not universal for simple hash map.

Exposing `Aggregator`, `Join` may be better than simple `hashtable` to users for easier integration.

jikunshang replied:
if so, `Cell` will be part of `hashtable` API, please call out that.

jikunshang commented:
please add `common` to cpplint tool, and fix potential lint issue. see https://github.com/intel/BDTK/pull/267
